File: crisscross/core_functions/slats.py
from collections import defaultdict
from colorama import Fore
import numpy as np


def get_slat_key(layer, slat_id):
    """
    Convenience function to generate slat key string.
    """
    return f'layer{layer}-slat{int(slat_id)}'


def convert_slat_array_into_slat_objects(slat_array):
    """
    Converts a slat array into a dictionary of slat objects for easy access.
    :param slat_array: 3D numpy array of slats - each point should either be a 0 (no slat) or a unique ID (slat here)
    :return: Dictionary of slats
    """
    slats = {}
    for layer in range(slat_array.shape[2]):
        for slat_id in np.unique(slat_array[:, :, layer]):
            if slat_id <= 0:  # removes any non-slat markers
                continue
            # slat coordinates are read in top-bottom, left-right
            slat_coords = np.argwhere(slat_array[:, :, layer] == slat_id).tolist()
            # generates a set of empty slats matching the design array
            slats[get_slat_key(layer + 1, int(slat_id))] = Slat(get_slat_key(layer + 1, int(slat_id)), layer + 1, slat_coords)
    return slats


class Slat:
    """
    Wrapper class to hold all of a slat's handles and related details.
    """
    def __init__(self, ID, layer, slat_coordinates, slat_length=32):
        self.ID = ID
        self.layer = layer
        self.max_length = slat_length
        self.reversed_slat = False  # flag to indicate if the slat has been reversed

        # converts coordinates on a 2d array to the handle number on the slat, and vice-versa
        self.slat_position_to_coordinate = {}
        self.slat_coordinate_to_position = {}
        if slat_coordinates != 'N/A':
            for index, coord in enumerate(slat_coordinates):
                self.slat_position_to_coordinate[index+1] = tuple(coord)
                self.slat_coordinate_to_position[tuple(coord)] = index + 1

        self.placeholder_list = []

        self.H2_handles = defaultdict(dict)
        self.H5_handles = defaultdict(dict)

    def reverse_direction(self):
        """
        Reverses the handle order on the slat (this should not affect the design placement of the slat).
        """
        new_slat_position_to_coordinate = {}
        new_slat_coordinate_to_position = {}
        for i in range(self.max_length):
            new_slat_position_to_coordinate[self.max_length - i] = self.slat_position_to_coordinate[i + 1]
            new_slat_coordinate_to_position[self.slat_position_to_coordinate[i + 1]] = self.max_length - i
        self.slat_position_to_coordinate = new_slat_position_to_coordinate
        self.slat_coordinate_to_position = new_slat_coordinate_to_position
        self.reversed_slat = not self.reversed_slat

    def get_sorted_handles(self, side='h2'):
        """
        Returns a sorted list of all handles on the slat (as they can be jumbled up sometimes, depending on the order they were created).
        :param side: h2 or h5
        :return: tuple of handle ID and handle dict contents
        """
        if side == 'h2':
            return sorted(self.H2_handles.items())
        elif side == 'h5':
            return sorted(self.H5_handles.items())
        else:
            raise RuntimeError('Wrong side specified (only h2 or h5 available)')

    def set_placeholder_handle(self, handle_id, slat_side, descriptor):
        """
        Assigns a placeholder to the slat, instead of a full handle.
        :param handle_id: Handle position on slat
        :param slat_side: H2 or H5
        :param descriptor: Description to use for placeholder
        :return: N/A
        """
        if handle_id < 1 or handle_id > self.max_length:
            raise RuntimeError('Handle ID out of range')

        if slat_side == 2:
            if handle_id in self.H2_handles:
                print(Fore.RED + 'WARNING: Overwriting handle %s, side 2 on slat %s' % (handle_id, self.ID))
            self.H2_handles[handle_id] = {'descriptor': descriptor}
        elif slat_side == 5:
            if handle_id in self.H5_handles:
                print(Fore.RED + 'WARNING: Overwriting handle %s, side 5 on slat %s' % (handle_id, self.ID))
            self.H5_handles[handle_id] = {'descriptor': descriptor}
        else:
            raise RuntimeError('Wrong slat side specified (only 2 or 5 available)')

        # placeholders are tracked here, for later replacement
        self.placeholder_list.append(f'handle-{handle_id}-h{slat_side}')

    def update_placeholder_handle(self,  handle_id, slat_side, sequence, well, plate_name, descriptor='No Desc.'):
        """
        Updates a placeholder handle with the actual handle.
        :param handle_id: Handle position on slat
        :param slat_side: H2 or H5
        :param sequence: Exact handle sequence
        :param well: Exact plate well
        :param plate_name: Exact plate name
        :param descriptor: Exact description of handle
        :return: N/A
        """

        input_id = f'handle-{handle_id}-h{slat_side}'
        if input_id not in self.placeholder_list:
            raise RuntimeError('Handle ID not found in placeholder list')
        else:
            self.placeholder_list.remove(input_id)

        if slat_side == 2:
            self.H2_handles[handle_id] = {'sequence': sequence, 'well': well, 'plate': plate_name,
                                          'descriptor': descriptor}
        elif slat_side == 5:
            self.H5_handles[handle_id] = {'sequence': sequence, 'well': well, 'plate': plate_name,
                                          'descriptor': descriptor}

    def set_handle(self, handle_id, slat_side, sequence, well, plate_name, descriptor='No Desc.'):
        """
        Defines the full details of a handle on a slat.
        :param handle_id: Handle position on slat
        :param slat_side: H2 or H5
        :param sequence: Exact handle sequence
        :param well: Exact plate well
        :param plate_name: Exact plate name
        :param descriptor: Exact description of handle
        :return: N/A
        """
        if handle_id < 1 or handle_id > self.max_length:
            raise RuntimeError('Handle ID out of range')
        if slat_side == 2:
            if handle_id in self.H2_handles:
                print(Fore.RED + 'WARNING: Overwriting handle %s, side 2 on slat %s' % (handle_id, self.ID))
            self.H2_handles[handle_id] = {'sequence': sequence, 'well': well, 'plate': plate_name,
                                          'descriptor': descriptor}
        elif slat_side == 5:
            if handle_id in self.H5_handles:
                print(Fore.RED + 'WARNING: Overwriting handle %s, side 5 on slat %s' % (handle_id, self.ID))
            self.H5_handles[handle_id] = {'sequence': sequence, 'well': well, 'plate': plate_name,
                                          'descriptor': descriptor}
        else:
            raise RuntimeError('Wrong slat side specified (only 2 or 5 available)')

File: flutter_app/lib/crisscross_core/slats.dart
/// Slat entity class representing DNA origami slats with handles.

import 'package:flutter/material.dart';


/// Rotates a point around a center in coordinate space by a given number of steps.
/// In 90deg mode, each step is 90deg CW (4 steps = full rotation).
/// In 60deg mode, each step is 60deg CW (6 steps = full rotation).
Offset rotateCoordinateSpace(Offset point, Offset center, int steps, String gridMode) {
  int modulus = gridMode == '90' ? 4 : 6;
  steps = steps % modulus;
  if (steps == 0) return point;

  double dx = point.dx - center.dx;
  double dy = point.dy - center.dy;

  for (int i = 0; i < steps; i++) {
    if (gridMode == '90') {
      // 90deg CW: (dx, dy) -> (dy, -dx)
      double newDx = dy;
      double newDy = -dx;
      dx = newDx;
      dy = newDy;
    } else {
      // 60deg CW: (dx, dy) -> (0.5*dx + 0.5*dy, -1.5*dx + 0.5*dy)
      double newDx = 0.5 * dx + 0.5 * dy;
      double newDy = -1.5 * dx + 0.5 * dy;
      dx = newDx.roundToDouble();
      dy = newDy.roundToDouble();
    }
  }

  return Offset(center.dx + dx, center.dy + dy);
}

Offset calculateCenter(Iterable<Offset> points) {
  double sumX = 0.0;
  double sumY = 0.0;
  int count = 0;

  for (final point in points) {
    sumX += point.dx;
    sumY += point.dy;
    count++;
  }

  if (count == 0) return Offset.zero;
  final inv = 1.0 / count;
  return Offset(sumX * inv, sumY * inv);
}

class Slat {
  /// Wrapper class to hold all of a slat's handles and related details.

  final String id;
  final String layer;
  final int maxLength;
  final int numericID;

  // Maps positions on the slat to coordinates on a 2D grid and vice-versa
  Map<int, Offset> slatPositionToCoordinate = {};
  Map<Offset, int> slatCoordinateToPosition = {};
  Offset centerCoordinate = Offset.zero;

  List<String> placeholderList = [];

  Map<int, Map<String, dynamic>> h2Handles = {};
  Map<int, Map<String, dynamic>> h5Handles = {};

  Color? uniqueColor;
  String slatType;
  String? phantomParent;

  Slat(this.numericID, this.id, this.layer, Map<int, Offset> slatCoordinates, {this.maxLength = 32, this.uniqueColor, this.slatType = 'tube', this.phantomParent}) {
      slatCoordinates.forEach((key, coord) {
        slatPositionToCoordinate[key] = coord;
        slatCoordinateToPosition[coord] = key;
      });
      centerCoordinate = calculateCenter(slatCoordinates.values.toList());
  }

  /// Sets a unique color for the slat, used for visualization.
  void setColor(Color color) {
    uniqueColor = color;
  }

  /// Clears the unique color of the slat.
  void clearColor() {
    uniqueColor = null;
  }

  /// Sets the ID of the reference slat this slat is associated with (or clear it).
  /// Setting this also implies this slat is a phantom slat.
  void setPhantom(String newPhantom) {
    phantomParent = newPhantom;
  }

  void updateCoordinates(Map<int, Offset> slatCoordinates){
    slatPositionToCoordinate = slatCoordinates;
    slatCoordinateToPosition = {for (var offset in slatCoordinates.entries) offset.value : offset.key};
    centerCoordinate = calculateCenter(slatCoordinates.values);
  }

  /// Reverses a slat, keeping all handles in the same physical position but changing their logical order.
  void reverseDirection() {
    Map<int, Offset> newSlatPositionToCoordinate = {};
    Map<Offset, int> newSlatCoordinateToPosition = {};

    // reverse all hande positions in the slat
    for (int i = 0; i < maxLength; i++) {
      newSlatPositionToCoordinate[maxLength - i] = slatPositionToCoordinate[i + 1]!;
      newSlatCoordinateToPosition[slatPositionToCoordinate[i + 1]!] = maxLength - i;
    }

    // updates h2 and h5 handles to match the new positions
    Map<int, Map<String, dynamic>> newH2Handles = {};
    Map<int, Map<String, dynamic>> newH5Handles = {};
    for (var entry in h2Handles.entries) {
      int newKey = maxLength - entry.key + 1;
      newH2Handles[newKey] = Map.from(entry.value);
    }

    for (var entry in h5Handles.entries) {
      int newKey = maxLength - entry.key + 1;
      newH5Handles[newKey] = Map.from(entry.value);
    }

    slatPositionToCoordinate = newSlatPositionToCoordinate;
    slatCoordinateToPosition = newSlatCoordinateToPosition;
    h2Handles = newH2Handles;
    h5Handles = newH5Handles;
  }

  /// Assigns a placeholder to the slat, instead of a full handle.
  void setPlaceholderHandle(int handleId, int slatSide, String value, String category) {
    // TODO: MORE STREAMLINED PLACEMENT LOGIC and REDUNDANCY/ERROR HANDLING
    if (handleId < 1 || handleId > maxLength) {
      throw Exception('Handle ID out of range');
    }

    if (slatSide == 2) {
      h2Handles[handleId] = {'value': value, 'category': category.toUpperCase(), 'placeholder': true};
    } else if (slatSide == 5) {
      h5Handles[handleId] = {'value': value, 'category': category.toUpperCase(), 'placeholder': true};
    } else {
      throw Exception('Wrong slat side specified (only 2 or 5 available)');
    }

    placeholderList.add('handle-$handleId-h$slatSide');
  }

  /// Updates a placeholder handle with the actual handle.
  void updatePlaceholderHandle(int handleId, int slatSide, String sequence, String well, String plateName, String value, String category, int concentration) {
    String inputId = 'handle-$handleId-h$slatSide';
    if (!placeholderList.contains(inputId)) {
      throw Exception('Handle ID not found in placeholder list');
    } else {
      placeholderList.remove(inputId);
    }

    if (slatSide == 2) {
      h2Handles[handleId] = {'sequence': sequence, 'well': well, 'plate': plateName, 'value': value, 'category': category.toUpperCase(), 'concentration': concentration};
    } else if (slatSide == 5) {
      h5Handles[handleId] = {'sequence': sequence, 'well': well, 'plate': plateName, 'value': value, 'category': category.toUpperCase(), 'concentration': concentration};
    }
  }

  /// Defines the full details of a handle on a slat.
  void setHandle(int handleId, int slatSide, String sequence, String well, String plateName, String value, String category, int concentration) {
    if (handleId < 1 || handleId > maxLength) {
      throw Exception('Handle ID out of range');
    }

    String inputId = 'handle-$handleId-h$slatSide';
    if (placeholderList.contains(inputId)) {
      placeholderList.remove(inputId);
    }

    if (slatSide == 2) {
      h2Handles[handleId] = {'sequence': sequence, 'well': well, 'plate': plateName, 'value': value, 'category': category.toUpperCase(), 'concentration': concentration};
    } else if (slatSide == 5) {
      h5Handles[handleId] = {'sequence': sequence, 'well': well, 'plate': plateName, 'value': value, 'category': category.toUpperCase(), 'concentration': concentration};
    } else {
      throw Exception('Wrong slat side specified (only 2 or 5 available)');
    }
  }

  bool checkPlaceholder(int handleID, int slatSide){
    return placeholderList.contains('handle-$handleID-h$slatSide');
  }

  /// Removes a handle from the slat at the given position and side.
  void removeHandle(int position, int side) {
    if (side == 2) {
      h2Handles.remove(position);
    } else if (side == 5) {
      h5Handles.remove(position);
    }
    placeholderList.remove('handle-$position-h$side');
  }

  /// Removes all handles from the slat.
  void clearAllHandles(){
    h2Handles.clear();
    h5Handles.clear();
    placeholderList.clear();
  }

  /// Removes all assembly handles from the slat.
  void clearAssemblyHandles() {
    // Collect keys to remove from h2Handles
    final keysToRemoveH2 = h2Handles.entries
        .where((entry) => entry.value['category'].contains('ASSEMBLY'))
        .map((entry) => entry.key)
        .toList();

    for (final key in keysToRemoveH2) {
      h2Handles.remove(key);
      final inputId = 'handle-$key-h2';
      placeholderList.remove(inputId);
    }

    // Collect keys to remove from h5Handles
    final keysToRemoveH5 = h5Handles.entries
        .where((entry) => entry.value['category'].contains('ASSEMBLY'))
        .map((entry) => entry.key)
        .toList();

    for (final key in keysToRemoveH5) {
      h5Handles.remove(key);
      final inputId = 'handle-$key-h5';
      placeholderList.remove(inputId);
    }
  }

  /// Calculates the molecular weight of the slat, based on the handles assigned.
  double getMolecularWeight() {
    int totalBases = 0;

    if (h2Handles.length < maxLength || h5Handles.length < maxLength) {
      throw Exception('Not all handles have been assigned on the slat, so the MW cannot be calculated yet.');
    }

    for (var handle in h2Handles.values) {
      if (handle['sequence'] == null) {
        throw Exception('Not all handles have been assigned on the slat, so the MW cannot be calculated yet.');
      }
      totalBases += (handle['sequence'].length as int);
    }
    for (var handle in h5Handles.values) {
      if (handle['sequence'] == null) {
        throw Exception('Not all handles have been assigned on the slat, so the MW cannot be calculated yet.');
      }
      totalBases += (handle['sequence'].length as int);
    }

    totalBases += 8064; // incorporating scaffold length
    totalBases += 5329; // incorporating length of core staples

    return totalBases * 327 - (totalBases - 1) * 18.015;
  }

  Slat copy() {
    final newSlat = Slat(numericID, id, layer, Map.from(slatPositionToCoordinate), maxLength: maxLength, uniqueColor: uniqueColor, slatType: slatType, phantomParent: phantomParent);

    newSlat.h2Handles = {
      for (var entry in h2Handles.entries)
        entry.key: Map.from(entry.value)
    };
    newSlat.h5Handles = {
      for (var entry in h5Handles.entries)
        entry.key: Map.from(entry.value)
    };

    newSlat.placeholderList = List.from(placeholderList);

    return newSlat;
  }

  void copyHandlesFromSlat(Slat slat){
    h2Handles = {
      for (var entry in slat.h2Handles.entries)
        entry.key: Map.from(entry.value)
    };
    h5Handles = {
      for (var entry in slat.h5Handles.entries)
        entry.key: Map.from(entry.value)
    };
    placeholderList = List.from(slat.placeholderList);
  }

}